File: src/main/scala/com/yogpc/qp/machines/workbench/IngredientWithCount.scala
package com.yogpc.qp.machines.workbench

import com.google.gson.{JsonArray, JsonElement, JsonObject}
import net.minecraft.item.ItemStack
import net.minecraft.item.crafting.Ingredient
import net.minecraft.network.PacketBuffer
import net.minecraft.util.JSONUtils
import net.minecraftforge.common.crafting.CraftingHelper

import scala.collection.immutable.ArraySeq
import scala.util.Try

case class IngredientWithCount(ingredient: Ingredient, count: Int) {

  def this(json: JsonObject) = {
    this(CraftingHelper.getIngredient(json), JSONUtils.getInt(json, "count"))
  }

  def this(stack: ItemStack) = {
    this(Ingredient.fromStacks(stack), stack.getCount)
  }

  def matches(stack: ItemStack): Boolean = {
    ingredient.test(stack) && stack.getCount >= count
  }

  def stackList: Seq[ItemStack] = {
    ArraySeq.unsafeWrapArray(ingredient.getMatchingStacks)
      .map(_.copy()).filter(s => !s.isEmpty).map { s =>
      s.setCount(count)
      s
    }
  }

  def shrink(stack: ItemStack): Boolean = {
    if (!matches(stack))
      return false
    stack.shrink(count)
    true
  }

  override def toString: String = {
    Try(ingredient.getMatchingStacks).toOption.flatMap(_.headOption) match {
      case Some(stack) => s"${stack.getItem} x$count}"
      case None => "Empty"
    }
  }

  def writeToBuffer(buffer: PacketBuffer): Unit = {
    ingredient.write(buffer)
    buffer.writeInt(count)
  }

  def serializeJson: JsonElement = {
    val obj = ingredient.serialize()
    obj match {
      case array: JsonArray => array.forEach(e => e.getAsJsonObject.addProperty("count", count))
      case jsonObject: JsonObject => jsonObject.addProperty("count", count)
    }
    obj
  }
}

object IngredientWithCount {
  private final val changer: JsonElement => JsonObject = _.getAsJsonObject

  def getSeq(json: JsonElement): Seq[IngredientWithCount] = {
    val factory: JsonObject => IngredientWithCount = new IngredientWithCount(_)
    if (json.isJsonArray) {
      import scala.jdk.CollectionConverters._
      json.getAsJsonArray.asScala.map(changer andThen factory).toSeq
    } else {
      Seq(factory(json.getAsJsonObject))
    }
  }

  def getSeq(stack: ItemStack): Seq[IngredientWithCount] = {
    Seq(new IngredientWithCount(stack))
  }

  def readFromBuffer(buffer: PacketBuffer): IngredientWithCount = {
    val ingredient = Ingredient.read(buffer)
    val count = buffer.readInt()
    IngredientWithCount(ingredient, count)
  }
}

File: src/main/scala/com/yogpc/qp/machines/workbench/IngredientWithCount.java
package com.yogpc.qp.machines.workbench;

import com.google.gson.JsonArray;
import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import net.minecraft.network.FriendlyByteBuf;
import net.minecraft.util.GsonHelper;
import net.minecraft.world.item.ItemStack;
import net.minecraft.world.item.crafting.Ingredient;
import net.minecraftforge.common.crafting.CraftingHelper;
import net.minecraftforge.items.ItemHandlerHelper;

import java.util.Arrays;
import java.util.List;
import java.util.function.Predicate;
import java.util.stream.StreamSupport;

public record IngredientWithCount(Ingredient ingredient, int count) implements Predicate<ItemStack> {
    public IngredientWithCount(JsonObject jsonObject) {
        this(CraftingHelper.getIngredient(modifyCount(jsonObject), false), GsonHelper.getAsInt(jsonObject, "count"));
    }

    public IngredientWithCount(ItemStack stack) {
        this(Ingredient.of(ItemHandlerHelper.copyStackWithSize(stack, 1)), stack.getCount());
    }

    @Override
    public boolean test(ItemStack stack) {
        return ingredient.test(stack) && stack.getCount() >= count;
    }

    boolean shrink(ItemStack stack) {
        if (test(stack)) {
            stack.shrink(count);
            return true;
        } else {
            return false;
        }
    }

    public List<ItemStack> stackList() {
        return Arrays.stream(ingredient.getItems())
                .filter(Predicate.not(ItemStack::isEmpty))
                .map(stack -> ItemHandlerHelper.copyStackWithSize(stack, count))
                .toList();
    }

    public JsonElement toJson() {
        var obj = ingredient.toJson();
        if (obj instanceof JsonArray jsonArray) {
            jsonArray.forEach(e -> e.getAsJsonObject().addProperty("count", count));
        } else if (obj instanceof JsonObject jsonObject) {
            jsonObject.addProperty("count", count);
        }
        return obj;
    }

    public void toPacket(FriendlyByteBuf buf) {
        ingredient.toNetwork(buf);
        buf.writeInt(count);
    }

    public static List<IngredientWithCount> getSeq(JsonElement jsonElement) {
        if (jsonElement instanceof JsonArray jsonArray) {
            return StreamSupport.stream(jsonArray.spliterator(), false)
                    .map(JsonElement::getAsJsonObject)
                    .map(IngredientWithCount::new)
                    .toList();
        } else {
            return List.of(new IngredientWithCount(jsonElement.getAsJsonObject()));
        }
    }

    public static IngredientWithCount fromPacket(FriendlyByteBuf buf) {
        var ingredient = Ingredient.fromNetwork(buf);
        var count = buf.readInt();
        return new IngredientWithCount(ingredient, count);
    }

    static JsonObject modifyCount(JsonObject object) {
        if (object.has("count")) {
            var clone = new JsonObject();
            object.entrySet().forEach(e -> clone.add(e.getKey(), e.getValue()));
            clone.remove("count");
            return clone;
        } else {
            return object;
        }
    }
}
